tests/bsim/bluetooth/mesh/src/test_beacon.c

```diff
@@ -908,13 +971,440 @@ static void test_rx_secure_beacon_interval(void)
+	provision(&tx_cfg);
+
+	sub = bt_mesh_subnet_get(0);
+	ASSERT_TRUE(sub != NULL);
+
+	ASSERT_TRUE(!atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_INITIATOR));
+	ASSERT_TRUE(!atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_IN_PROGRESS));
+	ASSERT_TRUE(!atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_PENDING));
+	ASSERT_TRUE(!atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_TEST));
+	ASSERT_TRUE(bt_mesh.iv_index == 0);
+
+	status = bt_mesh_subnet_kr_phase_get(0, &phase);
+	ASSERT_TRUE(status == STATUS_SUCCESS);
+	ASSERT_TRUE(phase == BT_MESH_KR_NORMAL);
+
+	/* Wait for SNB being advertised and switch beacon type between Beacon Intervals */
+	k_sleep(K_SECONDS(BEACON_INTERVAL + 5));
+
+	toggle_priv_beacon(tx_cfg.addr, 1);
+
+	bt_mesh_iv_update_test(true);
+	ASSERT_TRUE(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_TEST));
+
+	ASSERT_TRUE(bt_mesh_iv_update());
+	ASSERT_TRUE(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_IN_PROGRESS));
+	ASSERT_TRUE(bt_mesh.iv_index == 1);
+
+	/* Switch beacon type between Beacon Intervals */
+	k_sleep(K_SECONDS(BEACON_INTERVAL + 5));
+
+	toggle_priv_beacon(tx_cfg.addr, 0);
+
+	status = bt_mesh_subnet_update(BT_MESH_KEY_PRIMARY, net_key_new);
+	ASSERT_TRUE(status == STATUS_SUCCESS);
+	status = bt_mesh_subnet_kr_phase_get(BT_MESH_KEY_PRIMARY, &phase);
+	ASSERT_TRUE(status == STATUS_SUCCESS);
+	ASSERT_TRUE(phase == BT_MESH_KR_PHASE_1);
+
+	phase = BT_MESH_KR_PHASE_2;
+	status = bt_mesh_subnet_kr_phase_set(BT_MESH_KEY_PRIMARY, &phase);
+	ASSERT_TRUE(status == STATUS_SUCCESS);
+	status = bt_mesh_subnet_kr_phase_get(BT_MESH_KEY_PRIMARY, &phase);
+	ASSERT_TRUE(status == STATUS_SUCCESS);
+	ASSERT_TRUE(phase == BT_MESH_KR_PHASE_2);
+
+	k_sleep(K_SECONDS(BEACON_INTERVAL + 5));
+	toggle_priv_beacon(tx_cfg.addr, 1);
+
+	PASS();
+}
+
+static void test_rx_priv_interleave(void)
+{
+	int err;
+
+	bt_mesh_test_cfg_set(&rx_cfg, BEACON_INTERVAL_WAIT_TIME);
+	k_sem_init(&observer_sem, 0, 1);
+
+	err = bt_mesh_private_beacon_key(net_key, priv_beacon_key);
+	if (err) {
+		FAIL("Unable to generate beacon key");
+	}
+
+	err = bt_enable(NULL);
+	if (err) {
+		FAIL("Bluetooth init failed (err %d)", err);
+	}
+
+	expected_beacon = BEACON_TYPE_SECURE;
+	ASSERT_TRUE(wait_for_beacon(NULL, NULL));
+
+	expected_beacon = BEACON_TYPE_PRIVATE;
+	ASSERT_TRUE(wait_for_beacon(private_beacon_check, false));
+
+	/* IVU was started here */
+	ASSERT_TRUE(wait_for_beacon(private_beacon_check, false));
+	ASSERT_EQUAL(0x02, beacon.flags);
+	ASSERT_EQUAL(0x0001, beacon.iv_index);
+
+	memset(&beacon, 0, sizeof(beacon));
+	expected_beacon = BEACON_TYPE_SECURE;
+
+	ASSERT_TRUE(wait_for_beacon(NULL, NULL));
+	ASSERT_EQUAL(0x02, beacon.flags);
+	ASSERT_EQUAL(0x0001, beacon.iv_index);
+
+	/* KR was started here */
+	ASSERT_TRUE(wait_for_beacon(NULL, NULL));
+	ASSERT_EQUAL(0x03, beacon.flags);
+	ASSERT_EQUAL(0x0001, beacon.iv_index);
+
+	expected_beacon = BEACON_TYPE_PRIVATE;
+
+	err = bt_mesh_private_beacon_key(net_key_new, priv_beacon_key);
+
+	ASSERT_TRUE(wait_for_beacon(NULL, NULL));
```

Comment:
Should `private_beacon_check` be called here?

Reply:
Yeah, it can be :)